File: 0201-0300/0224-Basic Calculator/0224-Basic Calculator.py
class Solution:
    def calculate(self, s: str) -> int:
        St = []
        result = num = i = 0
        sign = 1
        while i < len(s):
            if s[i].isdigit():
                while i < len(s) and s[i].isdigit():
                    num = num * 10 + ord(s[i]) - ord('0')
                    i += 1
                result += sign * num
                num = 0
            else:
                if s[i] == '+':
                    sign = 1
                elif s[i] == '-':
                    sign = -1
                elif s[i] == '(':
                    St.append(sign)
                    St.append(result)
                    sign = 1
                    result = 0
                elif s[i] == ')':
                    result = St.pop() + St.pop() * result
                i += 1
        return result

File: 0201-0300/0224-Basic Calculator/0224-Basic Calculator.cpp
class Solution {
public:
    int calculate(string s) {
        stack<int> St;
        int result = 0, num = 0, i = 0;
        int sign = 1;
        while (i < s.size()) {
            if (isdigit(s[i])) {
                while (i < s.size() && isdigit(s[i])) {
                    num = num * 10 + (s[i++] - '0');
                }
                result += sign * num;
                num = 0;
            }
            else {
                if (s[i] == '+') {
                    sign = 1;
                }
                else if (s[i] == '-') {
                    sign = -1;
                }
                else if (s[i] == '(') {
                    St.push(sign);
                    St.push(result);
                    sign = 1;
                    result = 0;
                }
                else if (s[i] == ')') {
                    int prev = St.top();
                    St.pop();
                    sign = St.top();
                    St.pop();
                    result = prev + sign * result;
                }
                ++i;
            }
        }
        return result;
    }
};
